pkg/alertmanager/distributor/distributor.go

```diff
@@ -0,0 +1,317 @@
+
+	// Manager for subservices (Alertmanager Ring)
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	logger log.Logger
+
+	receivedRequests  *prometheus.CounterVec
+	amSends           *prometheus.CounterVec
+	amSendFailures    *prometheus.CounterVec
+	replicationFactor prometheus.Gauge
+}
+
+// New constructs a new Distributor
+func New(cfg Config, alertmanagersRing *ring.Ring, alertmanagerClientsPool AlertmanagerClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+		//replication:             alertmanagersRing.ReplicationFactor(),
+	}
+
+	d.initMetrics(reg)
+	d.replicationFactor.Set(float64(alertmanagersRing.ReplicationFactor()))
+
+	d.subservices, err = services.NewManager(alertmanagersRing)
+	if err != nil {
+		return nil, err
+	}
+	d.subservicesWatcher = services.NewFailureWatcher()
+	d.subservicesWatcher.WatchManager(d.subservices)
+
+	d.Service = services.NewBasicService(d.starting, d.running, d.stopping)
+	return d, nil
+}
+
+func (d *Distributor) initMetrics(r prometheus.Registerer) {
+	d.receivedRequests = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_received_requests_total",
+		Help: "The total number of requests received.",
+	}, []string{"user"})
+	d.amSends = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_total",
+		Help: "The total number of requests sent to the alertmanager.",
+	}, []string{"ingester"})
+	d.amSendFailures = promauto.With(r).NewCounterVec(prometheus.CounterOpts{
+		Name: "cortex_alertmanager_distributor_alertmanager_send_failures_total",
+		Help: "The total number of requests sent to the alertmanager that failed.",
+	}, []string{"ingester"})
+	d.replicationFactor = promauto.With(r).NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_alertmanager_distributor_replication_factor",
+		Help: "The configured replication factor.",
+	})
+}
+
+func (d *Distributor) starting(ctx context.Context) error {
+	return services.StartManagerAndAwaitHealthy(ctx, d.subservices)
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+func (d *Distributor) stopping(_ error) error {
+	return services.StopManagerAndAwaitStopped(context.Background(), d.subservices)
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+// TODO: This will go away in future after we gradually add support for the entire API.
```

Comment:
I would remove this TODO. Doesn't add much value. All code is subject to change by its nature ;)